src/balance.rs

```diff
@@ -26,41 +42,186 @@ impl<T: Config> From<Error<T>> for StatusCode {
 
 /// Balance adapter for providing basic access to balance cheques within the MoveVM.
 #[derive(Clone)]
-pub struct BalanceAdapter<T: Config> {
+pub struct BalanceAdapter<T: Config + SysConfig>
+where
+    AccountIdOf<T>: core::cmp::Eq + core::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     _pd_config: PhantomData<T>,
+    /// Copy and virtual cheques record of involved users.
+    cheques: Rc<RefCell<HashMap<AccountIdOf<T>, BalanceOf<T>>>>,
+    /// Copy of initial state, without tracking it.
+    initial_state: HashMap<AccountIdOf<T>, BalanceOf<T>>,
 }
 
-impl<T: Config> BalanceAdapter<T> {
+impl<T: Config + SysConfig> BalanceAdapter<T>
+where
+    AccountIdOf<T>: core::cmp::Eq + core::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     /// Create a new [`BalanceAdapter`].
     pub fn new() -> Self {
         BalanceAdapter {
             _pd_config: PhantomData,
+            cheques: Rc::new(RefCell::new(HashMap::new())),
+            initial_state: HashMap::new(),
+        }
+    }
+
+    /// Inserts an initial balance state, e.g. the executor of a script.
+    pub fn write_cheque(
+        &mut self,
+        account: &AccountIdOf<T>,
+        balance: &BalanceOf<T>,
+    ) -> DispatchResult {
+        // TODO (neutrinoks): Enable this feature to ensure, that balance is available!
+        // self.ensure_can_withdraw(account, balance)?;
```

Comment:
This would be a nice check! Can we uncomment it?

Reply:
Then, a couple of tests will fail. We have to add the new parameter in extrinsic `execute` and then we should be able to enable this again :blush: